sda/websocket.go

```diff
@@ -0,0 +1,366 @@
+		// Re-try to connect in case the websocket is just about to start
+		for a := 0; a < 10; a++ {
+			c.conn, _, err = d.Dial(fmt.Sprintf("ws://%s/%s/%s", url, c.service, path),
+				http.Header{})
+			if err == nil {
+				break
+			}
+			time.Sleep(100 * time.Millisecond)
+		}
+		if err != nil {
+			return nil, NewClientError(err)
+		}
+	}
+	defer func() {
+		if !c.keep {
+			c.Close()
+		}
+	}()
+	if err := c.conn.WriteMessage(websocket.BinaryMessage, buf); err != nil {
+		return nil, NewClientError(err)
+	}
+	_, rcv, err := c.conn.ReadMessage()
+	if err != nil {
+		return nil, NewClientError(err)
+	}
+	log.Lvlf4("Received %x", rcv)
+	return rcv, nil
+}
+
+// SendProtobuf wraps protobuf.(En|De)code over the Client.Send-function. It
+// takes the destination, a pointer to a msg-structure that will be
+// protobuf-encoded and sent over the websocket. If ret is non-nil, it
+// has to be a pointer to the struct that is sent back to the
+// client. If there is no error, the ret-structure is filled with the
+// data from the service. ClientError has a code and a msg in case
+// something went wrong.
+func (c *Client) SendProtobuf(dst *network.ServerIdentity, msg interface{}, ret interface{}) ClientError {
+	buf, err := protobuf.Encode(msg)
+	if err != nil {
+		return NewClientError(err)
+	}
+	path := strings.Split(reflect.TypeOf(msg).String(), ".")[1]
+	reply, cerr := c.Send(dst, path, buf)
+	if cerr != nil {
+		return NewClientError(cerr)
+	}
+	if ret != nil {
+		err := protobuf.DecodeWithConstructors(reply, ret,
+			network.DefaultConstructors(network.Suite))
+		return NewClientError(err)
+	}
+	return nil
+}
+
+// SendToAll sends a message to all ServerIdentities of the Roster and returns
+// all errors encountered concatenated together as a string.
+func (c *Client) SendToAll(dst *Roster, path string, buf []byte) ([][]byte, ClientError) {
+	msgs := make([][]byte, len(dst.List))
+	var errstrs []string
+	for i, e := range dst.List {
+		var err ClientError
+		msgs[i], err = c.Send(e, path, buf)
+		if err != nil {
+			errstrs = append(errstrs, fmt.Sprint(e.String(), err.Error()))
+		}
+	}
+	var err error
+	if len(errstrs) > 0 {
+		err = errors.New(strings.Join(errstrs, "\n"))
+	}
+	return msgs, NewClientError(err)
+}
+
+// Close sends a close-command to the connection.
+func (c *Client) Close() error {
+	if c.si != nil && c.conn != nil {
+		c.si = nil
+		return c.conn.Close()
+	}
+	return nil
+}
+
+// Pass the request to the websocket.
+type wsHandler struct {
+	ws      *WebSocket
+	service string
+	path    string
+}
+
+// Wrapper-function so that http.Requests get 'upgraded' to websockets
+// and handled correctly.
+func (t wsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	u := websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+		CheckOrigin: func(r *http.Request) bool {
+			return true
+		},
+		EnableCompression: true,
+	}
+	ws, err := u.Upgrade(w, r, http.Header{"Set-Cookie": {"sessionID=1234"}})
```

Comment:
? What does it mean ? It is not suposed to be "random" or stg ? I did not check the specs.